src/main/java/ch/cern/eam/wshub/core/services/administration/impl/UserGroupMenuServiceImpl.java

```diff
@@ -0,0 +1,321 @@
+
+        public GenericMenuEntry(MENU_Type menu) {
+            this.id = menu.getEXTMENUCODE();
+            this.description = menu.getMENUID().getMENUDESCRIPTION();
+            this.parent = menu.getEXTMENUPARENT();
+            this.functionId = menu.getMENUID().getMENUCODE();
+        }
+
+        public GenericMenuEntry(FUNCTION_Type function) {
+            this.id = function.getEXTMENUCODE();
+            this.description = function.getFUNCTIONID().getFUNCTIONDESCRIPTION();
+            this.parent = function.getEXTMENUPARENT();
+            this.functionId = function.getFUNCTIONID().getFUNCTIONCODE();
+        }
+
+        public String getId() {
+            return this.id;
+        }
+
+        public String getDescription() {
+            return this.description;
+        }
+
+        public String getFunctionId() {
+            return this.functionId;
+        }
+
+        public Object getParent() { return this.parent; }
+    }
+
+    private ExtMenusHierarchy getExtMenuHierarchy(InforContext context, MenuSpecification node) throws InforException {
+        MP6005_GetExtMenusHierarchy_001 getExtMenusHierarchy = new MP6005_GetExtMenusHierarchy_001();
+        getExtMenusHierarchy.setUSERGROUPID(new USERGROUPID_Type());
+        getExtMenusHierarchy.getUSERGROUPID().setUSERGROUPCODE(node.userGroup);
+
+        ExtMenusHierarchy result =
+                tools.performInforOperation(context, inforws::getExtMenusHierarchyOp, getExtMenusHierarchy)
+                        .getResultData().getExtMenusHierarchy();
+
+        return result;
+    }
+
+    private List<GenericMenuEntry> getExtMenuHierarchyAsList(InforContext context, MenuSpecification node) throws InforException {
+        ExtMenusHierarchy result = this.getExtMenuHierarchy(context, node);
+
+        List<GenericMenuEntry> menuEntries = new ArrayList<>();
+        for(MENU_Type menu : result.getMENU()) {
+            menuEntries.add(new GenericMenuEntry(menu));
+            menu.getFOLDER().stream().forEach(childFolder -> this.addFolderToEntries(menuEntries, childFolder));
+            menu.getFUNCTION().stream().forEach(childFunction -> this.addFunctionToEntries(menuEntries, childFunction));
+        }
+
+        return menuEntries;
+    }
+
+    private String decideMenuType(String menuCode, String[] path) {
+        String menuType;
+        if (menuCode == null || menuCode.isEmpty()) {
+            if (path.length <= 1) {
+                menuType = "M"; // Item is main menu
+            } else {
+                menuType = "F"; // Item is submenu
+            }
+        } else {
+            menuType = "S"; // Item is function
+        }
+
+        //WARN If two paths are identical, item will be added unreliably to one of them, as no item code is given (only path) to this function; we assume correct menu hierarchy already exists
+        return menuType;
+    }
+
+    private String[] calculateExistingPath(String[] words, List<GenericMenuEntry> menuEntries) {
```

Comment:
Java is much more List friendly and would even simplify some operations here

Reply:
Method deleted (long live the tree and getLevel).